L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+  // closure with true. The deadline here is an absolute time. The AlarmKey pointer 
+  // is an output parameter which can be used to cancel the alarm before it fires. 
+  // Note that the closure should not be blocking or long-running, as it would 
+  // block the worker thread in the EventManager from polling the file 
+  // descriptors. In case that the alarm is cancelled successfully in 
+  // ConsistentCancel, the closure will be run with false in the calling thread of 
+  // ConsistentCancel. Callers of ConsistentCancel need to be cautious to avoid 
+  // deadlock.
+  virtual void ScheduleAlarmIn(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, AlarmKey* key = nullptr) = 0;
+
+  // Adds a closure to the EventManager, which will be run at the earliest possible 
+  // time. The TaskKey pointer is an output parameter which can be used to cancel 
+  // the closure at a later time (currently not supported in 
+  // BaseEventManagerInterface). Note that the closure should not be blocking or 
+  // long-running, as it would block the worker thread in the EventManager from 
+  // polling the file descriptors.
+  virtual void RunFunction(std::function<void()> handler, Priority priority, TaskKey* key = nullptr) = 0;
+
+  struct Task {
+    std::function<void()> handler;
+    Priority priority;
+    TaskKey* key;
+    int internal_added_queue;
+  };
+  // Adds multiple closures to the event management system all at once.
+  virtual void RunFunctions(Task* tasks, int length) = 0;
+
+  // Cancels an alarm which has been added to the EventManager via 
+  // ScheduleAlarmIn(). This function will run the closure with false in the 
+  // calling thread if it successfully cancels the alarm.
+  virtual void ConsistentCancel(const AlarmKey& key) = 0;
+
+  // Cancels an alarm as ConsistentCancel(), but wait if the task is currently 
+  // running.
+  virtual void BlockingConsistentCancel(const AlarmKey& key) = 0;
+
+  // These two methods adjust a refcount of the EventManager, preventing it from 
+  // shutdown if the refcount is positive. ShutdownRef() should be called before 
+  // adding the first closure to the EventManager, and ShutdownUnref() can be 
+  // called after removing the last closure from the EventManager. Also, if you 
+  // need to keep the EventManager serving before receiving an RPC response, you 
+  // need to call ShutdownRef() before sending the RPC, and invoke ShutdownUnref() 
+  // when the response arrives.
+  virtual void ShutdownRef() = 0;
+  virtual void ShutdownUnref() = 0;
+};
+```
+
+The BaseEventManagerInterface is platform agnostic in order to accommodate the different polling models on different platforms (e.g. POSIX vs Windows).
+
+A specific interface EpollEventManagerInterface is derived from BaseEventManagerInterface for all the epoll-based polling models.
+
+```cpp
+// Represents a file descriptor on epoll-capable platforms.
+class EpollDescriptorInterface {
```

Comment:
I don't think an IO completion queue interface is within the scope of this design. Completion queues may need to use the eventmanager API.

Reply:
If an IOCP is out of scope for this design, why is an epoll one in scope? Aren't both just ways to implement the event manager API? 😉

Reply:
Is the plan to use libuv to implement this interface on Windows? If so, let's capture that. If not, I'd like to see at least a sketch of how this could be implemented on Windows. I assume such an implementation would build atop IO completions ports.

(There was an earlier comment here that I've since deleted. I hadn't fully digested the libuv parts yet.)

Reply:
Libuv will be implementing `LibuvEventManagerInterface` below. I'm not that familiar with libuv but I don't think libuv should implement this interface.

Reply:
"this interface" wasn't clear. I meant the `BaseEventManagerInterface`. Comment updated above.

It looks like the plan is to use libuv for all the platforms and add native implementations as needed. If that's the case, I suggest the following revisions to the paragraph before the libuv interface (additions in **bold**, removals <strike>struck through</strike>):

> Note that there will be no native EM implementation provided by gRPC for <strike>the epoll based interface</strike> **any platform** in the first step. Instead, gRPC will provide a default event manager implementation based on libuv. The libuv EM implementation will be <strike>applied to</strike> **used by** all the gRPC languages based on gRPC core, and the users may choose to override it with their custom implementation of the EM interface. gRPC may release more native event manager implementations down the road map.

Reply:
Thanks for the paragraph. I don't really see why we should state that no native EM implementation is provided for any platform. The libuv EM implementation will be a native one provided by gRPC, and it will be provided for most of the platforms. Maybe I misunderstood your question.

Reply:
Oh I see what you mean. When you say "native" you probably mean native for the target platform and I was more talking about "native" for gRPC.

I think I will rephrase it a little bit. Probably just say "Note that in the first step gRPC will provide a default EventManager implementation based on libuv. The libuv EM implementation will be...".

Reply:
That phrasing looks good to me.